docs/tutorials/basic/data.md

```diff
@@ -192,17 +128,390 @@ class SimpleIter:
+
+The *data* variables are called __free__ variables in the MXNet Symbol land that needs to bound with data. Learn more about [Symbol](http://mxnet.io/tutorials/basic/symbol.html).  
+
+We will feed the data iterator into the training problem using the `module` API. Learn more about [Module](http://mxnet.io/tutorials/basic/module.html).
+
+
+```python
+import logging
+logging.basicConfig(level=logging.INFO)
+
+n = 32
+data_iter = SimpleIter(['data'], [(n, 100)], 
+                  [lambda s: np.random.uniform(-1, 1, s)],
+                  ['softmax_label'], [(n,)], 
+                  [lambda s: np.random.randint(0, num_classes, s)])
+
+mod = mx.mod.Module(symbol=net)
+mod.fit(data_iter, num_epoch=5)
+```
+
+    INFO:root:Epoch[0] Train-accuracy=0.096875
+    INFO:root:Epoch[0] Time cost=0.281
+    INFO:root:Epoch[1] Train-accuracy=0.087500
+    INFO:root:Epoch[1] Time cost=0.008
+    INFO:root:Epoch[2] Train-accuracy=0.100000
+    INFO:root:Epoch[2] Time cost=0.008
+    INFO:root:Epoch[3] Train-accuracy=0.109375
+    INFO:root:Epoch[3] Time cost=0.008
+    INFO:root:Epoch[4] Train-accuracy=0.103125
+    INFO:root:Epoch[4] Time cost=0.012
+
+
+
+## Record IO
+Record IO is a file format used by *MXNet* for data IO. It compactly packs the data for efficient read and writes from various filesystems including distributed file systems like Hadoop HDFS and AWS S3.
+You can learn more about the thinking behind the design of `RecordIO` [here](http://mxnet.io/architecture/note_data_loading.html).
+
+*MXNet* provides __`MXRecordIO`__ and __`MXIndexedRecordIO`__ for sequential access of data and random access of the data.
+
+### MXRecordIO
+First let's look at an example on how to read and write sequentially using `mx.recordio.MXRecordIO`. The files are named with a `.rec` extension.
+
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'w')
+for i in range(5):
+    record.write('record_%d'%i)
+record.close()
+```
+
+We can read the data back by opening the file with a option `r` as below:
+
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'r')
+while True:
+    item = record.read()
+    if not item:
+        break
+    print (item)
+record.close()
+```
+
+    record_0
+    record_1
+    record_2
+    record_3
+    record_4
+
+
+
+### MXIndexedRecordIO
+`MXIndexedRecordIO` supports random or indexed access to the data. We will create a indexed record file and a corresponding index file as below:
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'w')
+for i in range(5):
+    record.write_idx(i, 'record_%d'%i)
+record.close()
+```
+
+Now we can access the individual records using the keys
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'r')
+record.read_idx(3)
+```
+
+
+
+
+    'record_3'
+
+
+
+You can also list all the keys in the file.
+
+
+```python
+record.keys
+```
+
+
+
+
```

Comment:
this is was exported from a notebook, not sure why this added many empty lines. removed these.